partition_snapshot_reader.hh

```diff
@@ -26,22 +26,46 @@
 #include "clustering_key_filter.hh"
 #include <boost/range/algorithm/heap_algorithm.hpp>
 
-template <typename Accounter>
+template <bool Reversing, typename Accounter>
 class partition_snapshot_flat_reader : public flat_mutation_reader::impl, public Accounter {
-    struct rows_position {
-        mutation_partition::rows_type::const_iterator _position;
-        mutation_partition::rows_type::const_iterator _end;
+    template<bool> struct rows_position_;
+    template<> struct rows_position_<false> {
+        using iter_type = mutation_partition::rows_type::const_iterator;
+        iter_type _position, _end;
     };
+    template<> struct rows_position_<true> {
+        using iter_type = mutation_partition::rows_type::const_reverse_iterator;
+        // The field order is reversed on purpose.
+        iter_type _end, _position;
+    };
+    using rows_position = rows_position_<Reversing>;
+    using rows_iter_type = typename rows_position::iter_type;
+
+    static rows_iter_type make_iterator(mutation_partition::rows_type::const_iterator it) {
+        if constexpr (Reversing)
+            return std::make_reverse_iterator(it);
+        else
+            return it;
+    }
 
     class heap_compare {
         position_in_partition::less_compare _less;
     public:
         explicit heap_compare(const schema& s) : _less(s) { }
         bool operator()(const rows_position& a, const rows_position& b) {
-            return _less(b._position->position(), a._position->position());
+            if constexpr (Reversing) {
+                // Here `reversed()` doesn't change anything, but keep it for consistency.
+                return _less(b._position->position().reversed(), a._position->position().reversed());
+            } else {
+                return _less(b._position->position(), a._position->position());
+            }
         }
         bool operator()(const range_tombstone_list::iterator_range& a, const range_tombstone_list::iterator_range& b) {
-            return _less(b.front().position(), a.front().position());
+            if constexpr (Reversing) {
+                return _less(b.back().end_position().reversed(), a.back().end_position().reversed());
+            } else {
+                return _less(b.front().position(), a.front().position());
+            }
```

Comment:
Alernatively, we could have seeded `_less` with a reversed comparator. But I see you also flip front/back, so maybe we're not ready for it.

Reply:
We couldn't have - it implicitly assumes that the schema used for comparator stays the same, otherwise UB occurs.